### ISSUE TITLE
Segmentations not rendered in OHIF-Viewer

### ISSUE THREAD

**Issue by ctorti** [OHIF/Viewers#2212]
Hello,

I have an intermittent issue that I've not been able to get to the bottom of.  I'm hoping that someone out there may offer some clues.

I'm copying and manipulating DICOM SEGs in [Pydicom](https://pydicom.github.io/) - an example workflow:

1. Read in a DICOM SEG ("Source")
2. Parse PixelData and manipulate the pixel array
3. Create a copy of the original SEG, replace PixelData with the modified pixel array (after converting to bytes)
4. Modify other tags within the new SEG so that the modified PixelData relate to a different set of DICOM images ("Target"), generate new SOPInstanceUID, etc.

After I've generated the new SEG file I perform some error checking (to verify that that the steps in 4 were correct). I wrote a function that parses PixelData from a SEG and plots ([Matplotlib](https://matplotlib.org/)) all frames in the pixel array, frame-by-frame.  The final step is to import Target SEG in the OHIF-Viewer and see if the segmentations are properly rendered by the viewer for the Target DICOM series.

I've managed to successfully do the above for a number of different datasets (DICOMs + DICOM SEG) that test the above under various use cases (e.g. Source and Target with the same/different Rows/Columns/PixelSpacing/SliceThickness/ImageOrientationPatient/etc.) but there are a couple of combinations that result in SEGs that when imported into the OHIF-Viewer do not display any segmentations at all. This despite my plotting function finding frames and displaying them as expected.

Not only that, but also the fact that there is nothing novel about the combinations that don't work. That is, the datasets used and processing steps made, are used/made for other use cases that do work.

I recognize that this is a rather complex issue involving interaction with a 3rd party package (Pydicom) that has nothing to do with the OHIF-Viewer, but I'm hoping that someone may have come across this issue before (and didn't post it here), may have some ideas as to what might be going wrong, or suggestions on where to look for clues.

Many thanks.

**Comment by pieper** [comment]
Note cross post in pydicom: https://github.com/pydicom/pydicom/issues/1290

**Comment by pieper** [comment]
I agree with @mrbean-bremen that example data would be required to help you debug.

**Comment by fedorov** [comment]
@ctorti did you run the [dciodvfy](http://dclunie.com/dicom3tools/dciodvfy.html) on the files you produced? Did you try using your DICOM SEG with independent tools that support it?
* https://github.com/QIICR/dcmqi (also used from 3D Slicer and MITK)
* https://highdicom.readthedocs.io/

**Comment by ctorti** [comment]
Thanks @mrbean-bremen and @pieper .

I can share the anonymised publically available series of DICOMs that I'm working with which originate from the [ACRIN](https://wiki.cancerimagingarchive.net/pages/viewpage.action?pageId=33948305) project at TheCancerImagingArchive.

But first, it's worth pointing out that I have some additional info that is probably relevant.  After importing the new SEG into the OHIF-Viewer the browser's java console reports the error:

"Single frame instance of multiframe class not supported"
![20201218_OHIF-Viewer_error_Single_frame_instance_of_multiframe_class_not_supported](https://user-images.githubusercontent.com/26715524/102774558-21191200-4383-11eb-9504-a60183251007.jpg)

It's been pointed out to me that this is a known issue affecting older versions of the OHIF-Viewer (I was running v2.1.0) with XNAT 1.7.6.5.  I was advised to upgrade to OHIF-Viewer v3.0.0, which requires XNAT v1.8.0.  I've recently upgraded both but have run into issues with XNAT, so I've been unable to verify whether the upgrade has resolved with original issue or not.

I'm going to try to get my XNAT working again and update this thread once I have new info to share.


And now the orginal and modified data with working and non-working examples:

I've shared both an example case that does work:  "Case3A-i" and one that doesn't: "Case3B-i".  In both of these cases the Source and Target DICOM series are the same:  [Source = Series no. 9](https://www.dropbox.com/sh/a91n68a0lvabh7u/AADQojMXXeOaNV_mQ2meDmkLa?dl=0) from a particular MR imaging session for a particular subject within ACRIN, and [Target = Series no. 3](https://www.dropbox.com/sh/y8u3qv0tuojkie7/AAC59WG19yJjMR2mYGHkw522a?dl=0) from the same MR imaging session.

The Source and Target DICOMs have:

```
   same Frame of Reference

   different image sizes:
       Source: (208, 256, 50) 
       Target: (378, 448, 21)

   different voxel sizes:
       Source: (0.8984375, 0.8984375, 2.9999990463256836) 
       Target: (0.51339286565781, 0.51339286565781, 7.499998569488525)

   different slice thickness:
       Source: 3.0 
       Target: 5.0

   different origins:
       Source: (-103.26176452636719, -153.43377685546875, -10.299073219299316) 
       Target: (-106.90052032470703, -153.20162963867188, -11.472214698791504)

   same patient orientation
       Source/Target: [0.9972648738498597, 1.8634260454120088e-10, 0.07391056342109259,
                       -0.022262997957353512, 0.9535561337886633, 0.3003915089278791,
                       -0.07047787104598309, -0.301215370979001, 0.9509480374756598]
```

Note that in addition to the direction cosines along rows and columns, I have appended the cross product of the former (hence why I report 9 elements in my patient orientation).

I created segmentations for both series in the OHIF-Viewer.  So I refer to the SEG file created for Series 9 as the [Source SEG](https://www.dropbox.com/s/lr49g7u16rsedb2/MR4_S9_tumour_and_brain_SEG_20201105_115939.dcm?dl=0), and the one for Series 3 as the [Target SEG](https://www.dropbox.com/s/mt8efmio99h6hg2/MR4_S3_tumour_and_brain_SEG_20201105_090359.dcm?dl=0).  

Both of these SEGs have two segments - one with a number of hand-drawn segmentations of the outer brain region, and the other some segmentations of a tumour.  For Case 3A-i, the objective was to copy the segmentation that exists for slice no. 25 from Source SEG to a new Target SEG such that the segmentation overlays onto slice no. 20 in Series 3.  

The resulting [new Target SEG](https://www.dropbox.com/s/scvsz928tapu8tt/20201215_164639_Case3A-i_SEG_from_s25_in_S9_to_s20_in_S3?dl=0) was created by modifying the original Target SEG accordingly.   In other words I am effectively copying-and-pasting a ROI without preserving the physical coordinates of the voxels that define the ROI.

For Case3B-i the objective was to copy the segmentation corresponding to slice no. 26 from Source SEG and copying it to Target SEG such that the spatial relationships are preserved.  

Due to the different image grids for Source and Target the Source segmentation that was to be copied was resampled to match the grid spacing of Target.  I hope that's enough info to be clear.

**Comment by ctorti** [comment]
It's worth pointing out the following threads I've just come across:

https://groups.google.com/g/xnat_discussion/c/KHqs3w4Vuvk/m/BgGrKNc_BgAJ

https://groups.google.com/g/xnat_discussion/c/5Lv1d-ztGP0/m/H9WoA2w4CgAJ

**Comment by ctorti** [comment]
> 
> 
> @ctorti did you run the [dciodvfy](http://dclunie.com/dicom3tools/dciodvfy.html) on the files you produced? Did you try using your DICOM SEG with independent tools that support it?
> 
>     * https://github.com/QIICR/dcmqi (also used from 3D Slicer and MITK)
> 
>     * https://highdicom.readthedocs.io/

Thanks @fedorov for your suggestions.

I've checked four SEG files using dciodvfy:  "SourceSeg", "TargetSeg", "WorkingNewTargetSeg" and "NonWorkingNewTargetSeg".

I get a number of warnings and an error for SourceSeg and TargetSeg but I will assume that since those SEGs were created within the OHIF-Viewer that they are not critical:

```
Warning - Value dubious for this VR - (0x0010,0x0010) PN Patient's Name  PN [0] = <ACRIN-FMISO-Brain-011> - Retired Person Name form
Segmentation
Warning - Coding Scheme Designator is deprecated - attribute <CodingSchemeDesignator> = <SRT> {<-- listed 4 times}
Error - ReferencedSeriesSequence present but Instance does not reference Instances - attribute <ReferencedSeriesSequence>
Warning - Attribute is not present in standard DICOM IOD - (0x0028,0x1052) DS Rescale Intercept
Warning - Attribute is not present in standard DICOM IOD - (0x0028,0x1053) DS Rescale Slope
Warning - Attribute is not present in standard DICOM IOD - (0x0028,0x1054) LO Rescale Type
Warning - Attribute is not present in standard DICOM IOD - (0x0028,0x9145) SQ Pixel Value Transformation Sequence
Warning - Dicom dataset contains attributes not present in standard DICOM IOD - this is a Standard Extended SOP Class
```

WorkingNewTargetSeg and NonWorkingNewTargetSeg has the above warnings and error plus:

```
Warning - Missing attribute or value that would be needed to build DICOMDIR - Study ID
Error - Value invalid for this VR - (0x0018,0x0050) DS Slice Thickness  DS [0] = <7.499998569488525> - Length invalid for this VR = 17, expected <= 16
Error - Value invalid for this VR - (0x0018,0x0088) DS Spacing Between Slices  DS [0] = <7.499998569488525> - Length invalid for this VR = 17, expected <= 16
Error - Dicom dataset contains invalid data values for Value Representations
```

I checked the value of StudyID for SourceSeg and TargetSeg and they are both 'No Study ID', whereas both WorkingNewTargetSeg and NonWorkingTargetSeg have empty strings.  I doubt that this would result in rendering issues for the OHIF-Viewer but since StudyID is required, I've modified my code so that the StudyID tag is left unchanged going forwards.

As for the next two errors, I've made a change in my code so that the character string values of SliceThickness and SpacingBetweenSlices is limited to 16 characters.  Again I'm doubtful that this was the cause of the issue, not least because it didn't prevent WorkingNewTargetSeg from being imported and rendered in the viewer.

The last error message seems to relate to the above two.

I've generated new Target SEGs and will see if the changes have made any difference.  Unfortunately I've broken my OHIF-Viewer so need to get that working again.  I'll report back once I have some news.

Thanks!

**Comment by pieper** [comment]
Thanks for all the detailed information, but it's hard to follow all the steps.  I see you have some dropbox links in the descriptions could you share the whole folder and include a readme about which files are which?  Also I'm not sure I follow the description in your first post - can you include the pydicom script also (e.g. in a gist or original repo?) and provide info about what you are trying to accomplish?

**Comment by ctorti** [comment]
Thanks @pieper for your suggestion and offer of support.  

I recently succeeded in upgrading a fresh Vagrant 1.7.6 XNAT to 1.8.0 (rather than my existing Vagrant 1.7.5).  With XNAT 1.8.0 I was able to import the problematic SEGs into OHIF-Viewer 3.0.0 and they were rendered correctly.  It appears therefore that the rendering issue I was having with some single-framed SEGs was an OHIF v2-related bug, as was reported/suggested by some others.  

I will now close this issue.
